lib/ansible/plugins/callback/foreman.py

```diff
@@ -77,157 +77,183 @@
+    """
+    return datetime.utcnow().isoformat()
+
+
 class CallbackModule(CallbackBase):
     CALLBACK_VERSION = 2.0
     CALLBACK_TYPE = 'notification'
     CALLBACK_NAME = 'foreman'
     CALLBACK_NEEDS_WHITELIST = True
 
-    FOREMAN_HEADERS = {
-        "Content-Type": "application/json",
-        "Accept": "application/json"
-    }
-    TIME_FORMAT = "%Y-%m-%d %H:%M:%S %f"
-
     def __init__(self):
         super(CallbackModule, self).__init__()
         self.items = defaultdict(list)
-        self.start_time = int(time.time())
+        self.facts = defaultdict(dict)
+        self.start_time = get_time()
 
     def set_options(self, task_keys=None, var_options=None, direct=None):
 
         super(CallbackModule, self).set_options(task_keys=task_keys, var_options=var_options, direct=direct)
 
-        self.FOREMAN_URL = self.get_option('url')
-        self.FOREMAN_SSL_CERT = (self.get_option('client_cert'), self.get_option('client_key'))
-        self.FOREMAN_SSL_VERIFY = str(self.get_option('verify_certs'))
-
-        self.ssl_verify = self._ssl_verify()
+        self.foreman_url = self.get_option('url')
+        ssl_cert = self.get_option('client_cert')
+        ssl_key = self.get_option('client_key')
 
         if HAS_REQUESTS:
-            requests_major = int(requests.__version__.split('.')[0])
-            if requests_major < 2:
-                self._disable_plugin('The `requests` python module is too old.')
+            requests_version = tuple(map(int, requests.__version__.split('.')))
+            if requests_version < (2, 14):
+                self._disable_plugin(u'The `requests` python module is older than 2.14.0.')
         else:
-            self._disable_plugin('The `requests` python module is not installed.')
+            self._disable_plugin(u'The `requests` python module is not installed.')
+
+        if self.foreman_url.startswith('https://'):
+            if not os.path.exists(ssl_cert):
+                self._disable_plugin(u'FOREMAN_SSL_CERT %s not found.' % ssl_cert)
 
-        if self.FOREMAN_URL.startswith('https://'):
-            if not os.path.exists(self.FOREMAN_SSL_CERT[0]):
-                self._disable_plugin('FOREMAN_SSL_CERT %s not found.' % self.FOREMAN_SSL_CERT[0])
+            if not os.path.exists(ssl_key):
+                self._disable_plugin(u'FOREMAN_SSL_KEY %s not found.' % ssl_key)
 
-            if not os.path.exists(self.FOREMAN_SSL_CERT[1]):
-                self._disable_plugin('FOREMAN_SSL_KEY %s not found.' % self.FOREMAN_SSL_CERT[1])
+        self.session = requests.Session()
+        self.session.verify = self._ssl_verify(str(self.get_option('verify_certs')))
+        self.session.cert = (ssl_cert, ssl_key)
 
     def _disable_plugin(self, msg):
         self.disabled = True
         if msg:
-            self._display.warning(msg + ' Disabling the Foreman callback plugin.')
+            self._display.warning(to_text(msg) + u' Disabling the Foreman callback plugin.')
         else:
```

Comment:
Note, the advice to use to_text() here was incorrect.  to_text should be used if the msg is not text where the msg enters the system as a whole.  It should not be used where msg is used (unless it is a non-string type until then.  But with a name like msg, it should be a string type when passed in here)

Reply:
So it should be `self._display.warning(msg + u' ...')`?

Reply:
Yes.  And then you look at where _display_plugin is called to make sure that msg is text type (unicode on python2; str on python3).  If there are variables instead of literal strings at that call site, you go back further to see if those variables are text type.  (Note that once you trace the variable back to an Ansible API, it should be text type unless it is explicitly documented as returning byte strings.  If it is an Ansible API and it does not return text type, that's a bug in the Ansible API).

Reply:
So yeah, the callers that I see are fine.  You should be able to simply remove to_text() here.

Reply:
The part of my comment about "unless it is a non-string type until then" means that you would use to_text if you have something like an exception object and were trying to pass that to display.warning.  Something like this:
``` python
try:
    1/0
except ZeroDivisionError as e:
    display.warning(u'Problem in code.  Error: ' + to_text(e))
```

In that case you are transforming a ZeroDivisionError object into a string.  That's an approptriate place to use to_text().